### PR TITLE
feat: add main unit tests for compose helper util

### PR DESCRIPTION
**Pull request by tushar5526**
Related to #4

### CODE REVIEW

File: server/utils.py [hunk 1]
```diff
@@ -53,6 +45,7 @@ class ComposeHelper:
 
     def __init__(self, compose_file_location: str, load_compose_file=True):
         self._compose_file_location = compose_file_location
+        self._compose = None
         if load_compose_file:
             self._compose = load_yaml_file(self._compose_file_location)
```
Comment (jatinkrmalik) [inline server/utils.py:50]: nit: Can be simplified to be much cleaner.

```python
def __init__(self, compose_file_location: str, load_compose_file=True):
    self._compose_file_location = compose_file_location
    self._compose = load_yaml_file(self._compose_file_location) if load_compose_file else None
```

File: server/utils.py [hunk 2]
```diff
@@ -110,6 +103,9 @@ def _generate_processed_compose_file(
             "services"
         ]["nginx"]
 
+        self._write_compose_file()
+
+    def _write_compose_file(self):
         with open(self._compose_file_location, "w") as yaml_file:
```
Comment (jatinkrmalik) [inline server/utils.py:109]: This can cause an `exception`, right? Shouldn't this be in `try/catch`?

File: tests/test_utils.py
Comment (jatinkrmalik) [inline tests/test_utils.py:1]: nit: Can improve cov.

- Additional tests could be written to cover error cases, such as what happens when an invalid compose file is provided, or if there are file access issues in `ComposeHelper`.

- Tests for the `_write_compose_file` method in `ComposeHelper` to ensure it correctly writes the YAML file.

File: tests/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        if (
+            "ports" in compose_helper._compose["services"][service]
+            and service != f"nginx_{deployment_namespace}"
+        ):
+            assert False, "Ports mapping present in processed compose file"
+        if "container_name" in compose_helper._compose["services"][service]:
+            assert False, "Container Name present in processed compose file"
+        if "restart" not in compose_helper._compose["services"][service]:
+            assert False, "Restart clause missing in processed compose file"
+        assert compose_helper._compose["services"][service]["restart"] == "always"
```
Comment (jatinkrmalik) [inline tests/test_utils.py:51]: Can make it much more readable by removing unnecessary nesting by extracting services from `compose_helper...`

```python
services = compose_helper._compose["services"]

    for service_name, service_config in services.items():
        is_nginx_service = service_name == f"nginx_{deployment_namespace}"
        
        assert "ports" not in service_config or is_nginx_service, \
            f"Ports mapping should not be present in {service_name}"
            
        assert "container_name" not in service_config, \
            f"Container Name should not be present in {service_name}"
            
        assert "restart" in service_config, \
            f"Restart clause missing in {service_name}"
            
        assert service_config["restart"] == "always", \
            f"Incorrect restart policy in {service_name}"
```

File: .github/workflows/test.yml
Comment (jatinkrmalik) [inline .github/workflows/test.yml:1]: nit: Since your dep list is small, we can ignore for now, but it's always a good practice to cache the dependencies to improve the performance and runtime of your workflow.

GitHub Actions provides a way to cache dependencies so that they don't have to be reinstalled on every run. We use it all the time.

```yaml
- name: Cache Python dependencies
  uses: actions/cache@v2
  with:
    path: ~/.cache/pip
    key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements*.txt') }}
    restore-keys: |
      ${{ runner.os }}-pip-
```

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    - name: Set up Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: 3.11  # Choose your Python version
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -r requirements.txt 
+        pip install -r requirements-dev.txt
```
Comment (jatinkrmalik) [inline .github/workflows/test.yml:24]: nit: Can separate steps for dep installation to improve clarity and simpler debugging.

```yaml
    - name: Install main dependencies
      run: |
        python -m pip install --upgrade pip
        pip install -r requirements.txt

    - name: Install development dependencies
      run: pip install -r requirements-dev.txt
```

**Review by jatinkrmalik — CHANGES_REQUESTED** [review]
Have suggested some enhancements! Most of the code is good, but we can work on `nit` to improve readability and maintanability! :)

**Comment by tushar5526** [comment]
Thanks for the review @jatinkrmalik I will make the changes soon!

File: server/utils.py (hunk 2 above)
Comment (tushar5526) [reply server/utils.py:109]: Yes, a lot of error handling is missing. I intended for the action to fail - but I should probably catch these errors and send a proper status code with an error message rather than GitHub Action receiving a 500 error 🤔

File: tests/test_utils.py (hunk 3 above)
Comment (tushar5526) [reply tests/test_utils.py:51]: Really helpful, thanks!

File: tests/test_utils.py
Comment (tushar5526) [reply tests/test_utils.py:1]: > Additional tests could be written to cover error cases, such as what happens when an invalid compose file is provided, or if there are file access issues in ComposeHelper.

Agreed, I will add some tests for negative cases as well. 

> Tests for the _write_compose_file method in ComposeHelper to ensure it correctly writes the YAML file.

How should one approach testing functions like these, they are just wrappers over library functions (`open`, `yaml`)?

I can think of mocking `yaml` and `open` library functions and checking whether the passed parameters are correct 🤔?

File: .github/workflows/test.yml
Comment (tushar5526) [reply .github/workflows/test.yml:1]: Thanks for the snippet as well.

File: server/utils.py (hunk 2 above)
Comment (jatinkrmalik) [reply server/utils.py:107]: Yep, some kind of handling will be nice! I would suggest having a custom exception so your logs are cleaner, like: `FileOpenException`.

File: tests/test_utils.py
Comment (jatinkrmalik) [reply tests/test_utils.py:1]: Yeah, mocking them should be good enough here. But if you want you can always write test data and validate that by comparing it to a fixture with that test data present. 

That might be overkill for now, unless you are doing something custom and async, you may not need that.

Comment (tushar5526) [reply tests/test_utils.py:1]: Done, thanks!

File: server/utils.py (hunk 2 above)
Comment (tushar5526) [reply server/utils.py:107]: Yes, I should have custom exceptions. Thanks, I will do the whole exception refactoring in a separate PR.
